src/node_trace_events.cc

```diff
@@ -0,0 +1,172 @@
+    return nullptr;
+  }
+
+  Utf8Value category(env->isolate(), categoryValue);
+  // If the value already exists in the set, insertion.first will point
+  // to the existing value. Thus, this will maintain a long lived pointer
+  // to the category c-string.
+  auto insertion = categoryGroups.insert(category.out());
+
+  // The returned insertion is a pair whose first item is the object that was
+  // inserted or that blocked the insertion and second item is a boolean
+  // indicating whether it was inserted.
+  return insertion.first->c_str();
+}
+
+static inline char GetPhase(Environment* env, const Local<Value>& arg) {
+  if (!arg->IsNumber()) {
+    env->ThrowTypeError("trace_events type must be a number.");
+    return 0;
+  }
+  Local<Context> context = env->isolate()->GetCurrentContext();
+  return static_cast<char>(arg->Int32Value(context).ToChecked());
+}
+
+static inline const char* GetName(Environment* env, const Local<Value>& arg) {
+  if (!arg->IsString()) {
+    env->ThrowTypeError("trace_events name must be a string.");
+    return nullptr;
+  }
+  Utf8Value nameValue(env->isolate(), arg);
+  return nameValue.out();
+}
+
+static inline int64_t GetId(Environment* env, const Local<Value>& arg, int64_t& id) {
+  if (arg->IsUndefined() || arg->IsNull()) {
+    id = 0;
+    return true;
+  }
+  else if (arg->IsNumber()) {
+    // This value may be zero, if no id is used
+    Local<Context> context = env->isolate()->GetCurrentContext();
+    id = arg->IntegerValue(context).ToChecked();
+    return true;
+  }
+  else {
+    env->ThrowTypeError("trace_events id must be a non-negative number.");
+    return false;
+  }
+}
+
+static void Emit(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  // Args: [type, category, name, id, argName, argValue]
+  CHECK_GE(args.Length(), 3);
+
+  // Check the category group first, to avoid doing more work if it's not
+  // enabled.
+  const char* category_group = GetCategoryGroup(env, args[1]);
+  const uint8_t* category_group_enabled = GetCategoryGroupEnabled(category_group);
+  if (*category_group_enabled == 0) return;
+
+  char phase = GetPhase(env, args[0]);
+  if (phase == 0) return;
+
+  const char* name = GetName(env, args[2]);
+  if (name == nullptr) return;
+
+  int64_t id = 0;
+  if (args.Length() >= 4 && !GetId(env, args[3], id)) return;
+
+  // TODO(@AndreasMadsen): Two extra arguments are not supported.
+  // TODO(@AndreasMadsen): String values are not supported.
+  int32_t num_args = 0;
+  const char* arg_names[1];
+  uint8_t arg_types[1];
+  uint64_t arg_values[1];
+
+  if (args.Length() < 6 || (args[4]->IsUndefined() && args[5]->IsUndefined())) {
+    num_args = 0;
+  }
+  else if (args[4]->IsString() && args[5]->IsNumber()) {
+    num_args = 1;
+    Local<Context> context = env->isolate()->GetCurrentContext();
+    printf("name [A]: %s\n", name); // REQUEST
+    arg_names[0] = Utf8Value(env->isolate(), args[4]).out();
```

Comment:
@addaleax could you take a look? I don't understand why fetching the c-string of `args[4]` changes the value of `name`.

Reply:
@AndreasMadsen The memory associated with `Utf8Value.out()` is tied to the lifetime of the `Utf8Value` itself – you can’t really use `GetName` like this, sorry.

(It sounds like what happens here is that when `name = GetName()` is called, the name string is converted to utf8, that memory is returned by `GetName()`, then freed because the `Utf8Value` goes out of scope; when you try to update `arg_names[0]` here, the same thing happens, and this just re-uses the memory allocated for `name`.)

Reply:
Thanks, this does make sense to me. However, I'm not sure how to appropriately fix it. Can you help me?

Reply:
The easiest way would be to just keep the `Utf8Value` around while you use the string (basically what you do [here](https://github.com/nodejs/node/pull/15538/files/35c21fb62a840efbecec0c8a498eb10d86570003#diff-5a0ddabd197fdff8886e9fb9d85ff024R34) – this is also what we do in most of the codebase, so grepping around should show enough examples).

The second-easiest thing would probably be to create `std::string`s from those `Utf8Values` and use them instead, e.g. `GetName()` could return a `std::string` and should start working :)

Reply:
Okay. I was worried if the `if`-block scoping would cause some issues.

Reply:
Oh – right, the block scope is an issue here. I think you can actually move the `Utf8Value` declaration to outside of the diff … if that doesn’t work, `std::string` is probably the way to go